### DISCUSSION TITLE
Understand seperation of data in DEV and PROD?

### DISCUSSION THREAD

**Discussion by gerhardmpl** [Budibase/budibase#5442] [Help]
First, let me apologize if my testing and/or reasoning is wrong, as I may be completely off the mark here.

While preparing a PoC app for test in production, I wanted to understand how data in the production (_PROD_) environment (aka "published app") is protected and separated from data in the development (_DEV_) environment (aka "unpublished, work in progess app". I am aware that the datasource is shared and that there is only one budibase stack / budibase database). My approach was to test this with a simple app on the BB cloud offering, where I changed data in the published (_PROD_) and changed (_DEV_) app to see what was happening respectively.

1. initial app publish from _DEV_ to _PROD_ includes dev data (let's call them initial dev or `idev` data)
2. changes on `idev` data in _PROD_ make it to _DEV_
3. changes on `idev` data in _DEV_ do not make it to _PROD_
4. deleting `idev` data in _PROD_ does not delete data in _DEV_
5. adding new data in _PROD_ (`nprod`) makes it to _DEV_
6. changes on `nprod` data in _PROD_ make it to _DEV_
7. changes on `nprod` data in _DEV_ do not make it to _PROD_
8. deleting `nprod` data in _PROD_ does not delete data in _DEV_
9. adding new data in _DEV_ (`ndev`) does not make it to _PROD_
10. changes on `ndev` data in _DEV_ do not make it to _PROD_
11. doing some minor changes to screens in _DEV_ and republishing the app brings all _DEV_ data to _PROD_

Maybe I am misunderstanding the deployment process, or my tests are not adequate, but I have some issues with my findings.

Depending on the app data, having _PROD_ data in _DEV_ will most certainly be a compliance issue in a corporate environment because of segregation of duty and need to know principle.

Another problem is that I would expect a lot of crash and burn or smoke tests in _DEV_, which will create, change, and delete a lot of arbitrary data. This must not be published to _PROD_ when the app is (re-) published.

This leaves me wondering how an app can be used in _PROD_ while development continues although we cannot have different datasources for _DEV_ and _PROD_? How can we protect production data in an agile development process with short sprints and regular app releases?

As I said: as a BB noob, I may be completely off the mark here, but I would love to understand the deployment process and the mandatory separation of _DEV_ and _PROD_ data.

edit: some clarification and typos

**Comment by mjashanks** [comment]
Hey @gerhardmpl 

All of your observations are correct. Prod replicates to dev, but not vise versa, until you publish. On publish, all data gets replicated from dev to prod.

One point I would clarify...
> the datasource is shared and that there is only one budibase stack / budibase database

Budibase dev and production databases are actually 2 separate databases that sync with each other at the various points that you describe.

I completely understand your issue around compliance, and agree - I would like to see a time where we have more segregation over prod and dev data.

Our decision to implement it this way was the most simple solution. A Budibase app database (both dev and prod) contains both
1. Schema for the apps - Data schemas, Design section and Automations definitions.
2. Data of internal tables

When we publish from dev to prod, we need to **at least** replicate the schema for the apps. Copying ONLY the schema is completely possible, and is a feature that we should definitely consider in the future. 

However, copying the data is a little more complicated, as sometimes you would want to copy the data, and sometimes you might not. An example of when you do want to copy data is when updating lookup or reference data.

The ideal scenario here is to give the user control of what data to copy over. As you can imagine, presenting this to the user gets quite complicated. This is why we haven't implemented it (priorities).

So, we had to choose between
- Never sync the data on publish
- Always sync the data on publish

Since we had no other way to copy data across, we chose to "Always sync the data"

I realise that this is not the ideal solution, but hopefully, you can understand the decision-making process behind the current method.

**Reply by gerhardmpl** [reply]
@mjashanks thank you for your detailed explanation. It helps me to better understand Budibase’s options of use and I appreciate that.

What I still would like to understand is how a best practice development and deployment process would look like that does not impact the data in PROD.

> Prod replicates to dev, but not vice versa, until you publish. On publish, all data gets replicated from dev to prod.

This means that you cannot test in _DEV_, because all data from _DEV_ is replicated to _PROD_? For this reason, I export the app to a separate _DEV_ instance and manually merge (edit) the changed apps schema (data schemas, design section, automations) in the _PROD_ instance. Personally, I find that tedious and error-prone, but now I understand that more clearly. Or is there a better approach?

Reading #3682 I came across an important hint: 

> Syncing happens whenever you open the app in the builder.

This is in line with my finding that you cannot just deploy to _PROD_ at any given time while you are working in the builder because that might cause data replication issues. There must be some sort of downtime for user and for builder log out / log in, syncing and deployment. But I need to test that more thoroughly and with more patience (which is not my core competency).

#2869 and #4630 are interesting reads as well.

**Reply by magickaito** [reply]
@gerhardmpl have you figured out a clean way to separate dev and production data? I just realised this challenge today...

**Reply by gerhardmpl** [reply]
@magickaito unfortunately there seems to be no good way to separate _DEV_ and _PROD_ data. For the moment, I am doing it the way I described above, since I do not (yet) trust syncing data when deploying to _PROD_.

**Comment by phifogg** [comment]
Oh ... wow. This is a great thread. I just wondered why some of my data is in sync and some not. I believe it is very very important to understand when a sync happens in which direction.

I started digging into this as I saw duplicate AutoID's one some records and wondered how/when this would happen. But now that makes more sense. I probably need to think about this more as I especially do not want test data to be published (or for the sake, be available on prod).
